fili-core/src/main/java/com/yahoo/bard/webservice/table/availability/PartitionAvailability.java

```diff
@@ -0,0 +1,110 @@
+// Copyright 2017 Yahoo Inc.
+// Licensed under the terms of the Apache license. Please see LICENSE.md file distributed with this work for terms.
+package com.yahoo.bard.webservice.table.availability;
+
+import com.yahoo.bard.webservice.data.config.names.TableName;
+import com.yahoo.bard.webservice.table.Column;
+import com.yahoo.bard.webservice.table.PhysicalTable;
+import com.yahoo.bard.webservice.table.resolver.DataSourceConstraint;
+import com.yahoo.bard.webservice.util.IntervalUtils;
+import com.yahoo.bard.webservice.util.SimplifiedIntervalList;
+
+import org.joda.time.Interval;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+import javax.validation.constraints.NotNull;
+
+/**
+ * An implementation of {@link com.yahoo.bard.webservice.table.availability.Availability}
+ * that unions different tables in druid that have the same columns together,
+ * in a way that the same column in two different table contains different values.
+ * Therefore if only values in one of the table is requested we only need to consider that table's availability.
+ */
+public class PartitionAvailability implements Availability {
+
+    private final Set<PhysicalTable> sourceTables;
+    private final Set<Column> columns;
+    private final Function<DataSourceConstraint, Set<PhysicalTable>> partitionFunction;
+
+    /**
+     * Constructor.
+     *
+     * @param sourceTables  The set of tables that have the same columns
+     * @param columns  The set of columns in concern of this availability(i.e. ParitionAvailability)
+     * @param partitionFunction  A funtion that transform
+     * a {@link com.yahoo.bard.webservice.table.resolver.DataSourceConstraint} to a set of
+     * {@link com.yahoo.bard.webservice.table.PhysicalTable}'s
+     */
+    public PartitionAvailability(
+            @NotNull Set<PhysicalTable> sourceTables,
+            @NotNull Set<Column> columns,
+            @NotNull Function<DataSourceConstraint, Set<PhysicalTable>> partitionFunction
+    ) {
+        this.sourceTables = sourceTables;
+        this.columns = columns;
+        this.partitionFunction = partitionFunction;
+    }
+
+    @Override
+    public Set<TableName> getDataSourceNames() {
+        return sourceTables.stream()
+                .map(PhysicalTable::getTableName)
+                .collect(Collectors.toSet());
+    }
+
+    @Override
+    public Map<Column, List<Interval>> getAllAvailableIntervals() {
+        return getColumnToIntervalsMapping().entrySet().stream()
+                .filter(entry -> columns.contains(entry.getKey()))
```

Comment:
I feels like we might be able to use a single stream, rather than dumping to an intermediate collection, applying this filter before we start to aggregate intervals.